p2p/net/swarm/dial_worker_test.go

```diff
@@ -342,3 +358,206 @@ func TestDialWorkerLoopConcurrentFailureStress(t *testing.T) {
+			}
+		}
+	}
+	go acceptAndIgnore(ch1, tcpL1)()
+	go acceptAndIgnore(ch2, tcpL2)()
+	go acceptAndIgnore(ch3, tcpL3)()
+
+	ranker := func(addrs []ma.Multiaddr) []*network.AddrDelay {
+		res := make([]*network.AddrDelay, 0)
+		for _, a := range addrs {
+			switch {
+			case a.Equal(silAddr1):
+				res = append(res, &network.AddrDelay{Addr: a, Delay: 0})
+			case a.Equal(silAddr2):
+				res = append(res, &network.AddrDelay{Addr: a, Delay: 1 * time.Second})
+			case a.Equal(tcpAddr):
+				res = append(res, &network.AddrDelay{Addr: a, Delay: 2 * time.Second})
+			case a.Equal(silAddr3):
+				res = append(res, &network.AddrDelay{Addr: a, Delay: 3 * time.Second})
+			default:
+				t.Errorf("unexpected address %s", a)
+			}
+		}
+		return res
+	}
+
+	// should connect to quic with both tcp and quic address
+	s1.dialRanker = ranker
+	s2addrs := []ma.Multiaddr{tcpAddr, silAddr1, silAddr2, silAddr3}
+	s1.Peerstore().AddAddrs(s2.LocalPeer(), s2addrs, peerstore.PermanentAddrTTL)
+	reqch := make(chan dialRequest)
+	resch := make(chan dialResponse)
+	worker1 := newDialWorker(s1, s2.LocalPeer(), reqch)
+	go worker1.loop()
+	defer worker1.wg.Wait()
+
+	reqch <- dialRequest{ctx: context.Background(), resch: resch}
+	select {
+	case <-ch1:
+	case <-time.After(1 * time.Second):
+		t.Fatal("expected dial to tcp1")
+	case <-resch:
+		t.Fatalf("didn't expect connection to succeed")
+	}
+	select {
+	case <-ch2:
+	case <-time.After(2 * time.Second):
+		t.Fatalf("expected dial to tcp2")
+	case <-resch:
+		t.Fatalf("didn't expect connection to succeed")
+	}
+	select {
+	case res := <-resch:
+		if !res.conn.RemoteMultiaddr().Equal(tcpAddr) {
+			log.Errorf("invalid connection address. expected %s got %s", tcpAddr, res.conn.RemoteMultiaddr())
+		}
+	case <-time.After(2 * time.Second):
+		t.Fatalf("expected dial to succeed")
+	}
+	close(reqch)
+	s1.ClosePeer(s2.LocalPeer())
+	s1.peers.ClearAddrs(s2.LocalPeer())
+	select {
+	case <-ch3:
+		t.Errorf("didn't expect tcp call")
+	case <-time.After(2 * time.Second):
+	}
+
+	quicFirstRanker := func(addrs []ma.Multiaddr) []*network.AddrDelay {
+		m := make([]*network.AddrDelay, 0)
+		for _, a := range addrs {
+			if _, err := a.ValueForProtocol(ma.P_TCP); err == nil {
+				m = append(m, &network.AddrDelay{Addr: a, Delay: 500 * time.Millisecond})
+			} else {
+				m = append(m, &network.AddrDelay{Addr: a, Delay: 0})
+			}
+		}
+		return m
+	}
+
+	// tcp should connect after delay
+	s1.dialRanker = quicFirstRanker
+	s2.ListenClose(quicAddr)
+	s1.Peerstore().AddAddrs(s2.LocalPeer(), []ma.Multiaddr{quicAddr, tcpAddr}, peerstore.PermanentAddrTTL)
+	reqch = make(chan dialRequest)
+	resch = make(chan dialResponse)
+	worker2 := newDialWorker(s1, s2.LocalPeer(), reqch)
+	go worker2.loop()
+	defer worker2.wg.Wait()
+
+	reqch <- dialRequest{ctx: context.Background(), resch: resch}
+	select {
+	case res := <-resch:
+		t.Fatalf("expected a delay before connecting %s", res.conn.LocalMultiaddr())
+	case <-time.After(400 * time.Millisecond):
```

Comment:
would be nice to use a mock clock for all these times. There's a very simple mockclock implementation in `core/test/mockclock` that avoids some of the issues in https://github.com/benbjohnson/clock. Here's an interface it can fulfill: https://github.com/libp2p/go-libp2p/blob/master/p2p/host/autorelay/options.go#L174. It uses a slightly different interface to a normal clock, but it fits with what you're doing here.

Reply:
using this now. thanks!